File: solutions/13/main.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"hmcalister/AdventOfCode/clawmachine"
	"log/slog"
	"os"
	"regexp"
	"runtime/pprof"
	"strconv"
	"time"
)

const (
	CPU_PROFILE_FILEPATH string = "profile"
)

func main() {
	debugFlag := flag.Bool("debug", false, "Debug Flag")
	inputFilePath := flag.String("inputFile", "puzzleInput", "Path to input file.")
	selectedPart := flag.Int("part", 0, "Part to execute. Must be 1 or 2.")
	profile := flag.Bool("profile", false, "Flag to profile program")
	flag.Parse()
	if *profile {
		f, err := os.Create(CPU_PROFILE_FILEPATH)
		if err != nil {
			slog.Error("could not create cpu profile file", "file", CPU_PROFILE_FILEPATH)
			os.Exit(1)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	logFileHandler := SetLogging(*debugFlag)
	defer logFileHandler.Close()

	inputFile, err := os.Open(*inputFilePath)
	if err != nil {
		slog.Error("error opening input file", "error", err)
		os.Exit(1)
	}
	defer inputFile.Close()

	fileScanner := bufio.NewScanner(inputFile)
	if err != nil {
		slog.Error("error processing file input", "error", err)
		os.Exit(1)
	}

	var result int
	computationStartTime := time.Now()
	switch *selectedPart {
	case 1:
		result, err = Part01(fileScanner)
	case 2:
		result, err = Part02(fileScanner)
	default:
		slog.Error("invalid part selected, part must be one of 1 or 2", "part selected", *selectedPart)
		os.Exit(1)
	}
	computationEndTime := time.Now()
	if err != nil {
		slog.Error("error encountered during computation", "error", err, "part selected", *selectedPart)
		os.Exit(1)
	}

	slog.Info("computation completed", "result", result, "computation time elapsed (ns)", computationEndTime.Sub(computationStartTime).Nanoseconds())
}

func parseInputToClawMachines(fileScanner *bufio.Scanner) []*clawmachine.ClawMachine {
	machines := make([]*clawmachine.ClawMachine, 0)

	buttonARegex := regexp.MustCompile(`Button A: X([+-]\d*), Y([+-]\d*)`)
	buttonBRegex := regexp.MustCompile(`Button B: X([+-]\d*), Y([+-]\d*)`)
	prizeRegex := regexp.MustCompile(`Prize: X=([+-]?\d*), Y=([+-]?\d*)`)

	// We still have some input
	for fileScanner.Scan() {
		buttonALine := fileScanner.Text()
		if !fileScanner.Scan() {
			slog.Debug("unexpected EOF when reading button B line")
			return machines
		}
		buttonBLine := fileScanner.Text()
		if !fileScanner.Scan() {
			slog.Debug("unexpected EOF when reading prize line")
			return machines
		}
		prizeLine := fileScanner.Text()
		fileScanner.Scan()

		slog.Debug("next claw machine input read", "button A line", buttonALine, "button B line", buttonBLine, "prize line", prizeLine)

		buttonAMatches := buttonARegex.FindStringSubmatch(buttonALine)
		if buttonAMatches == nil || len(buttonAMatches) != 3 {
			slog.Error("button A line not of expected format", "button A line", buttonALine, "button A regex matches", buttonAMatches)
			continue
		}
		buttonAX, errA := strconv.ParseFloat(buttonAMatches[1], 64)
		buttonAY, errB := strconv.ParseFloat(buttonAMatches[2], 64)
		if errors.Join(errA, errB) != nil {
			slog.Error("could not parse button A line", "button A line", buttonALine, "button A regex matches", buttonAMatches)
			continue
		}

		buttonBMatches := buttonBRegex.FindStringSubmatch(buttonBLine)
		if buttonBMatches == nil || len(buttonBMatches) != 3 {
			slog.Error("button B line not of expected format", "button B line", buttonBLine, "button B regex matches", buttonBMatches)
			continue
		}
		buttonBX, errA := strconv.ParseFloat(buttonBMatches[1], 64)
		buttonBY, errB := strconv.ParseFloat(buttonBMatches[2], 64)
		if errors.Join(errA, errB) != nil {
			slog.Error("could not parse button B line", "button B line", buttonBLine, "button B regex matches", buttonBMatches)
			continue
		}

		prizeMatches := prizeRegex.FindStringSubmatch(prizeLine)
		if prizeMatches == nil || len(prizeMatches) != 3 {
			slog.Error("prize line not of expected format", "prize line", prizeLine, "prize regex matches", prizeMatches)
			continue
		}
		prizeX, errA := strconv.ParseFloat(prizeMatches[1], 64)
		prizeY, errB := strconv.ParseFloat(prizeMatches[2], 64)
		if errors.Join(errA, errB) != nil {
			slog.Error("could not parse prize line", "prize line", prizeLine, "prize regex matches", prizeMatches)
		}

		nextClawMachine := clawmachine.NewClawMachine(
			buttonAX,
			buttonAY,
			buttonBX,
			buttonBY,
			prizeX,
			prizeY,
		)
		slog.Debug("input parsed to claw machine", "claw machine", nextClawMachine)
		machines = append(machines, nextClawMachine)
	}

	return machines
}

func Part01(fileScanner *bufio.Scanner) (int, error) {
	clawMachines := parseInputToClawMachines(fileScanner)
	totalCost := 0
	validMachines := 0
	for _, machine := range clawMachines {
		cost, err := machine.ComputeLowestTokenCost()
		if err != nil {
			slog.Error("error when computing token cost", "error", err)
			continue
		}
		totalCost += cost
		validMachines += 1
		slog.Debug("found lowest token cost", "current machine token cost", cost, "updated total token cost", totalCost)

	}
	slog.Debug("finished computing claw results", "total claw machines", len(clawMachines), "valid claw machines", validMachines)
	return totalCost, nil
}

func Part02(fileScanner *bufio.Scanner) (int, error) {
	clawMachines := parseInputToClawMachines(fileScanner)
	totalCost := 0
	validMachines := 0
	for _, machine := range clawMachines {
		machine.FixUnitConversion()
		cost, err := machine.ComputeLowestTokenCost()
		if err != nil {
			slog.Error("error when computing token cost", "error", err)
			continue
		}
		totalCost += cost
		validMachines += 1
		slog.Debug("found lowest token cost", "current machine token cost", cost, "updated total token cost", totalCost)

	}
	slog.Debug("finished computing claw results", "total claw machines", len(clawMachines), "valid claw machines", validMachines)
	return totalCost, nil
}

File: solutions/13/src/main.rs
use std::error;
use std::fs::File;
use std::io::{BufRead, BufReader};

use clap::Parser;
use claw_machine::ClawMachine;
use nalgebra::{Matrix2, Vector2};
use regex::Regex;
use std::time::SystemTime;
#[allow(unused_imports)]
use tracing::{debug, error, info, trace};

mod claw_machine;
mod logging;

/// Program to solve Advent of Code puzzles
#[derive(Parser, Debug)]
#[command(version, about)]
struct CommandLineArgs {
    /// File to read input from
    #[arg(long, default_value = "puzzleInput")]
    input_file: String,

    /// Selected part of puzzle to solve
    #[arg(long, value_parser=clap::value_parser!(u8).range(1..=2))]
    part: u8,
}

fn main() {
    logging::set_logging();
    let args = CommandLineArgs::parse();
    trace!("command line args" = ?args, "parsed command line args");

    let input_file_handle = File::open(args.input_file).expect("could not open input file");
    let input_file_reader = BufReader::new(input_file_handle);

    let start_time = SystemTime::now();
    let computation_result = match args.part {
        1 => part01(input_file_reader),
        2 => part02(input_file_reader),
        _ => unreachable!(), // clap has filtered out all other possibilities.
    }
    .expect("computation did not produce a value");
    let elapsed_time = start_time.elapsed().unwrap();

    info!(
        "computation_result" = computation_result,
        "elapsed_time" = elapsed_time.as_nanos(),
        "computation complete"
    );
}

fn parse_input_to_claw_machines(
    mut input_file_reader: BufReader<File>,
) -> Result<Vec<claw_machine::ClawMachine>, Box<dyn error::Error>> {
    let mut claw_machines: Vec<ClawMachine> = Vec::new();
    let mut line = String::new();
    let button_a_regex = Regex::new(r"Button A: X([+-]?\d+), Y([+-]?\d+)")?;
    let button_b_regex = Regex::new(r"Button B: X([+-]?\d+), Y([+-]?\d+)")?;
    let prize_regex = Regex::new(r"Prize: X=([+-]?\d+), Y=([+-]?\d+)")?;

    loop {
        line.clear();
        input_file_reader.read_line(&mut line)?;
        let Some(button_a_captures) = button_a_regex.captures(&line) else {
            error!(?line, "could not match button a line");
            return Err(Box::from("could not parse button a line"));
        };
        let (_, [button_a_x_increment, button_a_y_increment]) = button_a_captures.extract();
        let button_a_x_increment = button_a_x_increment.parse::<f64>()?;
        let button_a_y_increment = button_a_y_increment.parse::<f64>()?;
        
        line.clear();
        input_file_reader.read_line(&mut line)?;
        let Some(button_b_captures) = button_b_regex.captures(&line) else {
            error!(?line, "could not match button b line");
            return Err(Box::from("could not parse button b line"));
        };
        let (_, [button_b_x_increment, button_b_y_increment]) = button_b_captures.extract();
        let button_b_x_increment = button_b_x_increment.parse::<f64>()?;
        let button_b_y_increment = button_b_y_increment.parse::<f64>()?;
        
        line.clear();
        input_file_reader.read_line(&mut line)?;
        let Some(prize_captures) = prize_regex.captures(&line) else {
            error!(?line, "could not match prize line");
            return Err(Box::from("could not parse prize line"));
        };
        let (_, [prize_x, prize_y]) = prize_captures.extract();
        let prize_x = prize_x.parse::<f64>()?;
        let prize_y = prize_y.parse::<f64>()?;

        let new_claw_machine = claw_machine::new(
            Matrix2::new(
                button_a_x_increment,
                button_b_x_increment,
                button_a_y_increment,
                button_b_y_increment,
            ),
            Vector2::new(prize_x, prize_y),
        );
        trace!("machine"=?new_claw_machine, "parsed claw machine");
        claw_machines.push(new_claw_machine);

        input_file_reader.read_line(&mut line)?;
        if input_file_reader.fill_buf()?.is_empty() {
            break;
        }
    }

    return Ok(claw_machines)
}

fn part01(input_file_reader: BufReader<File>) -> Option<i64> {
    let claw_machines = match parse_input_to_claw_machines(input_file_reader) {
        Ok(machines) => machines,
        Err(e) => {
            error!("error"=?e, "error occurred during input parsing");
            return None;
        }
    };
    trace!(?claw_machines, "parsed claw machines");

    let total_cost: i64 = claw_machines
        .iter()
        .filter_map(|machine| match machine.solve_machine() {
            Ok(machine_result) => {
                debug!(?machine, ?machine_result, "solved machine");
                Some(machine_result)
            },
            Err(e) => {
                trace!(?machine, "error"=?e, "could not solve machine");
                None
            }
        }).sum();

    Some(total_cost)
}

fn part02(input_file_reader: BufReader<File>) -> Option<i64> {
    for line_result in input_file_reader.lines() {
        let line = line_result.unwrap();
        debug!("line" = line, "read line from input file");
    }

    None
}
